[feat] make Rocq interruptible

gares opened the pull request:
TLDR: Rocq's interpreter runs in a thread that can be interrupted.

<img width="1198" height="465" alt="Screenshot From 2026-04-01 13-04-33" src="https://github.com/user-attachments/assets/180820a6-bc31-40ed-b8c4-18c01987218a" />
<img width="1198" height="465" alt="Screenshot From 2026-04-01 13-06-11" src="https://github.com/user-attachments/assets/5fe63eac-7e7d-4c98-a73c-e3b4418cbf20" />

### Technology

Requires sel 0.7 (for promises) and memprof-limits (battle tested in coq-lsp, kudos to @gadmm)

### Expected behavior
- in stop-on-first-error-mode it just stops
- in continue-on-error it continues (maybe this is not the best behavior, but this is how it is in this PR)

Editing the document interrupts execution as usual, but now also if the Rocq interpreter is looping. Hence I don't think there is a real need for a keybinding to interrupt Rocq. Thanks to this, continuous mode is usable if looping sentences are around.

### Other design choices

The proverThread has 3 ways to sequentialize:
- eventually_run: put in a queue and return a promise
- try_run: check if the thread is busy, and if not run now (with a timeout)
- run: if the thread is busy, interrupt and put in the queue the interrupted task, then run now (with a timeout)

This PR uses *eventually_run* for parsing and execution (checking) of sentences; *try_run* for queries (including hovering); *run* for he rest, including printing the proof view. Note that you can check two documents at the same time, vsrocq always did "time sharing" bu checking one sentence of each, so busy really means "busy on any open document". These choices are trivial to change in the code. The main tradeoff is that interrupting/restarting jobs loses time. If, for example, hovering does that, one may really slow down a long sentence (each time you move the mouse you risk to restart it).
 
## Architecture

There was a big issue, namely that Rocq is not thread safe (nor purely functional). Scenario: the execution manager (thread) is running on a sentence: it installed a global state and is possibly even modifying it in place before snapshotting it back; at the same time the lsp manager gets a request, like hover or query or completion, that currently installs a Rocq state and run some code.  Hence we reorganize the code as follows

<img width="1255" height="1427" alt="Screenshot From 2026-04-01 12-57-30" src="https://github.com/user-attachments/assets/f6d0751e-876a-413e-8c17-5321bb2433ac" />

All calls to Rocq APIs are either pure, or serialized by the prover thread. Queries are run on a best effort (if the thread is busy for too long they are aborted). Parsing/Execution is serialized and made interruptible by the user.
A test compares the call graph of documentManager with a reference, checking that each call is either whitelisted (pure code) or goes trough a safety barrier (makes the call sequential).

- [x] block on first error broken

gares commented:
CC @rtetley, when you have a minute, could you please add a stop button to the UI ?

gares commented:
After some thinking the plan could be:
- rework the source tree to use `(implicit_transitive_deps false)` in dune project and have all Rocq calls / types in a dedicated library `rocq_enclave`. All code outside the enclave should not list coq-core.* in the dune file, hence it cannot call Rocq APIs. In this way we precisely control what we call
- make sure that Rocq calls are made in a single thread and serialized one after the other
  - with an API to interrupt, possibly using memprof_limits
  - use the same priorities we use for sel in the API: if a call has higher priority interrupt the running one
  - bonus: if a call has higher priority interrupt the running one (after say 0.1) and run the high prio one, then restart the interrupted one

This includes all Rocq calls, the queries, the parser, etc

Some Rocq code may not need this, being pure, but it is not so clear how to know it and ensure it Rocq side. Eg `Search` seems pure, but prints the results using the current notations, so it is not really pure.

rtetley committed:
feat: add a button and command to send intterupt`

rtetley commented:
> CC @rtetley, when you have a minute, could you please add a stop button to the UI ?

Done



> Along with a stop button could we have a key binding to interrupt as well

I created a command that you can call from the palette. What default keybinding should we use ?

rtetley committed:
feat: add vscoq/interrupt verb to language server

WIP

rtetley committed:
fix: pass a documentUri to the interrupt notification to get state

Without a state we can't kill the current execution

Durbatuluk1701 commented:
I know it has been previously suggested that if you have an in-progress proof and input the "Step Backward" key bind it should be interpreted as "Interrupt", so that is a possibility.

I also think maybe given that we have commands bound to up, down, and right arrow keys, maybe left arrow key could be "Interrupt"?

gares committed:
Interruption via memprof-limits

gares committed:
Make e2e tests more reproducible

gares committed:
update doc

gares committed:
warnings

gares committed:
callgraph tool

gares committed:
cleanup ProverThread APIs

gares committed:
sequentialize document initialization

gares committed:
improve ProverThread API forcing exn printing to occur inside

gares committed:
sequentialize start_library

gares committed:
remove unneeded unfreeze_state call when interpreting to top

gares committed:
protect delegate proof termination

gares committed:
simplify higher order code in create_parse_event

gares committed:
do not use get_default_proof_using at a random state

gares committed:
sequentialize get proof

gares committed:
Run queries in Rocq thread

gares committed:
Make queries run now (with a timeout)

gares commented on language-server/protocol/extProtocol.ml:
```diff
@@ -76,6 +85,9 @@ module Notification = struct
       | "prover/interpretToEnd" ->
         let+ params = Lsp.Import.Json.message_params params InterpretToEndParams.t_of_yojson in
         InterpretToEnd params
+      | "vscoq/interrupt" ->
```
prover!

gares commented on client/src/manualChecking.ts:
```diff
@@ -1,19 +1,15 @@
 
+export const sendInterrupt = (editor: TextEditor, client: Client) => {
+    const textDocument = makeVersionedDocumentId(editor);
+    client.sendNotification("vscoq/interrupt", {textDocument});
```
prover

gares replied:
document extension

gares committed:
renaming

gares committed:
update doc

gares committed:
improve build log

gares committed:
update lock.json

gares committed:
rocq pre 9.2 is not interruption safe

gares committed:
print job names in debug logs

gares committed:
send the PV once the execution is over

since observe/execute is now a promise, being after it in the SEL
queue does not guarantee sequentiality

gares committed:
conservative approach: options to disable threading

gares committed:
proverThread: simplify log

gares committed:
send pv also on cached state

gares commented on client/package.json:
```diff
@@ -128,14 +128,20 @@
+            "scope": "workspace",
+            "type": "boolean",
+            "default": false,
+            "description": "Hovering and other queries preempt checking tasks (recommended Rocq >= 9.2). "
```
```suggestion
            "description": "Hovering and other queries preempt checking (recommended Rocq >= 9.2). "
```

gares committed:
Remove last preemption waiting for Rocq to sort out

https://github.com/rocq-prover/rocq/issues/22041

gares committed:
fixup rebase

gares committed:
ci

gares committed:
better logs

gares committed:
fix test

gares committed:
ci

gares committed:
fix bug in overview introduced in #1223

gares committed:
Forcibly send a debug message so that the output panel is created

gares committed:
ci
